src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1336 @@
+    value_array[nodata_mask] = (
+        retention_volume_array[nodata_mask] * replacement_cost)
+    return value_array
+
+
+def impervious_op(lulc_array, lulc_nodata, sorted_lucodes, 
+        impervious_lookup_array):
+    """Convert LULC array to a binary array where 1 is directly connected
+    impervious LULC type and 0 is not.
+
+    Args:
+        lulc_array (numpy.ndarray): 2D array of integer LULC codes
+        lulc_nodata (int): nodata value for the LULC array
+        sorted_lucodes (numpy.ndarray): a 1D array of the LULC codes in order 
+            from smallest to largest
+        impervious_lookup_array (numpy.ndarray): 1D boolean array indicating 
+            whether each LULC type is impervious or not. The LULC code at 
+            index ``i`` of ``sorted_lucodes`` is impervious iff 
+            ``impervious_lookup_array[i] == 1``.
+
+    Returns:
+        2D boolean numpy.ndarray of the same shape as ``lulc_array``
+    """
+    is_impervious_array = numpy.full(lulc_array.shape, NODATA)
+    valid_mask = (lulc_array != lulc_nodata)
+    # bin the lulc values such that 
+    # lulc_array[i,j] == sorted_lucodes[lulc_index[i,j]]
+    lulc_index = numpy.digitize(lulc_array, sorted_lucodes, right=True)
+    is_impervious_array[valid_mask] = (
+        impervious_lookup_array[lulc_index][valid_mask])
+    return is_impervious_array 
+
+
+def adjust_op(ratio_array, avg_ratio_array, near_impervious_lulc_array, 
+        near_road_array):
+    """Apply the retention ratio adjustment algorithm to an array of ratios. 
+    This is meant to be used with raster_calculator. Assumes that the nodata
+    value for all four input arrays is the global NODATA.
+
+    Args:
+        ratio_array (numpy.ndarray): 2D array of stormwater retention ratios
+        avg_ratio_array (numpy.ndarray): 2D array of averaged ratios
+        near_impervious_lulc_array (numpy.ndarray): 2D boolean array where 1 
+            means this pixel is near a directly-connected LULC area
+        near_road_array (numpy.ndarray): 2D boolean array where 1 
+            means this pixel is near a road centerline
+        
+    Returns:
+        2D numpy array of adjusted retention ratios. Has the same shape as 
+        ``retention_ratio_array``.
+    """
+    adjusted_ratio_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    adjustment_factor_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (ratio_array != NODATA) &
+        (avg_ratio_array != NODATA) &
+        (near_impervious_lulc_array != NODATA) &
+        (near_road_array != NODATA))
+
+    # adjustment factor:
+    # - 0 if any of the nearby pixels are impervious/connected;
+    # - average of nearby pixels, otherwise
+    is_impervious = (
+        near_impervious_lulc_array[valid_mask] | 
+        near_road_array[valid_mask])
+    adjustment_factor_array[valid_mask] = (avg_ratio_array[valid_mask] * 
+        (1 - is_impervious))
```

Comment:
Neat! Changed it to `bool` dtype. 
It could be condensed into one statement like so
```
    adjustment_factor_array[valid_mask] = (
        avg_ratio_array[valid_mask] * ~(
            near_impervious_lulc_array[valid_mask] | 
            near_road_array[valid_mask]
        ).astype(bool))
```
but the intent seems clearer as it is, defining `is_not_impervious` separately.